<center>
    
## <font color='maroon'>ASTR 20500</font>
## <font color='maroon'>homework assignment 2</font>
    
#### <font color='black'>distributed: Friday, October 8</font>
#### <font color='black'>due: Thursday, October 14 (11pm)</font>
    
    
#### 30 points total (+5 potential extra-credit points)
</center>


Information you need to carry out exercises in this assignment is available in the first two distributed notebooks: [02_probability](https://drive.google.com/file/d/18r-zCP4W67VW5HAOElvmpXkTB2Q_kX88/view?usp=sharing), [03_probability_distributions_poisson](https://drive.google.com/file/d/12vmgU44wLBNsaCbQ5EVjc3Ta0dMczupt/view?usp=sharing), and [04_expectation_variance_std](https://drive.google.com/file/d/1MgbxmGCYyk7TsQe0ATO1ZSzbMIWGt7-m/view?usp=sharing). 

Background Python information useful for this assignment is provided in the [hw02_background_info](https://drive.google.com/file/d/119-Zh3tKkoOQqIspQbfMBsHWJDKuVucK/view?usp=sharing) notebook. 

Although there are 6 exercises provided, default is 4 exercises in this assignment:  
   * **_choose exercises 2 or 6_** for the 2nd exercise. Exercise 2 is more straightforward and is intended for students who are new to programming, while exercise 6 is for those who already have experience with Python, working with lists, and know how implement functions with flexible number of arguments (there are opportunities for neat hacks here).  
   
   * **_choose exercise 4 or 5_**. These should be comparable in terms of difficulty. 
   

### <font color='blue'>Exercise 1: practicing programming with while loops
### <font color='blue'>    and exploring limitations of the 64-bit representation of floating point numbers (10 points)</font> 

**1a. (5 points)** Starting with 

    x = 1.0
    
write a [<tt>while loop</tt>](https://www.w3schools.com/python/python_while_loops.asp) that divides x by 2 at each iteration and prints: 1) number of the loop iteration and 2) value of x at each iteration. The loop must exit if x cannot be distinguished from 0. 
    
What is the value of <tt>x</tt> during the iteration before the loop exits (i.e., before the condition is met)?  (3 points) Explain briefly the value of $x$ at which condition <tt>x==0</tt> is satisfied (and only approximately, i.e. no detailed explanation is needed) (2 points). 

In [58]:
i=1
x=1.0
while x:
  print(i,x)
  x /= 2
  i += 1

1 1.0
2 0.5
3 0.25
4 0.125
5 0.0625
6 0.03125
7 0.015625
8 0.0078125
9 0.00390625
10 0.001953125
11 0.0009765625
12 0.00048828125
13 0.000244140625
14 0.0001220703125
15 6.103515625e-05
16 3.0517578125e-05
17 1.52587890625e-05
18 7.62939453125e-06
19 3.814697265625e-06
20 1.9073486328125e-06
21 9.5367431640625e-07
22 4.76837158203125e-07
23 2.384185791015625e-07
24 1.1920928955078125e-07
25 5.960464477539063e-08
26 2.9802322387695312e-08
27 1.4901161193847656e-08
28 7.450580596923828e-09
29 3.725290298461914e-09
30 1.862645149230957e-09
31 9.313225746154785e-10
32 4.656612873077393e-10
33 2.3283064365386963e-10
34 1.1641532182693481e-10
35 5.820766091346741e-11
36 2.9103830456733704e-11
37 1.4551915228366852e-11
38 7.275957614183426e-12
39 3.637978807091713e-12
40 1.8189894035458565e-12
41 9.094947017729282e-13
42 4.547473508864641e-13
43 2.2737367544323206e-13
44 1.1368683772161603e-13
45 5.684341886080802e-14
46 2.842170943040401e-14
47 1.4210854715202004e-14
48 7.105427357601002e-15

The value of $x$ during the iteration before the loop exits is $x=5\cdot 10^{-324}$ on iteration $i=1075$. It looks like the condition $x==0$ is satisfied when $x$ is on the order of $10^{-325}$, which is a pretty small number to say the least (pun intended)!

**1b. (5 points)** Use the loop similar to the loop above but which exits when $1+x$ becomes indistinguishable from 1. 
    
What is the value of <tt>x</tt> during the iteration before the loop exits (i.e., before the loop exit condition is satisfied)? (1 point) Explain briefly why the value of $x$ at which the loop exit condition is satisfied is so different from the value of <tt>x</tt> that satisfies the condition in 1a (4 points). 

In [59]:
1+1/2**50==1

False

In [60]:
i=1
x=1.0
while 1+x>1:
  print(i,x)
  x /= 2
  i += 1

1 1.0
2 0.5
3 0.25
4 0.125
5 0.0625
6 0.03125
7 0.015625
8 0.0078125
9 0.00390625
10 0.001953125
11 0.0009765625
12 0.00048828125
13 0.000244140625
14 0.0001220703125
15 6.103515625e-05
16 3.0517578125e-05
17 1.52587890625e-05
18 7.62939453125e-06
19 3.814697265625e-06
20 1.9073486328125e-06
21 9.5367431640625e-07
22 4.76837158203125e-07
23 2.384185791015625e-07
24 1.1920928955078125e-07
25 5.960464477539063e-08
26 2.9802322387695312e-08
27 1.4901161193847656e-08
28 7.450580596923828e-09
29 3.725290298461914e-09
30 1.862645149230957e-09
31 9.313225746154785e-10
32 4.656612873077393e-10
33 2.3283064365386963e-10
34 1.1641532182693481e-10
35 5.820766091346741e-11
36 2.9103830456733704e-11
37 1.4551915228366852e-11
38 7.275957614183426e-12
39 3.637978807091713e-12
40 1.8189894035458565e-12
41 9.094947017729282e-13
42 4.547473508864641e-13
43 2.2737367544323206e-13
44 1.1368683772161603e-13
45 5.684341886080802e-14
46 2.842170943040401e-14
47 1.4210854715202004e-14
48 7.105427357601002e-15

The value of $x$ during the iteration before the loop exits is $x=2.22\cdot 10^{-16}$ on iteration $i=53$. The value of $x$ at which the loop exit condition is satisfied here in part (b) is so different from the loop exit condition in part (a) because, unlike in part (a), the function we are testing is $1+x$, so for any value of $x\in[0,1]$, the value of the function $1+x$ will be of the form $1.XXXXX$ repeating, depending on the value of $x$, where $X$ is just an arbitrary placeholder. Since the value before the decimal point is predetermined, namely $1$, then the memory only has 52 additional bits to work with as it is dividing through our value of $x$. In part (a), since there was no value before the decimal point (or alternatively, it was 0), the computer was able to do some clever stuff with its memory by shifting over bits to add additional decimal places and the like, allowing us to have $1023+52=1075$ bits total to use.

### <font size=4 color='blue'>Exercise 2: implementing a function to compute the sum of list entries (7 points)</font> 

Implement a function <tt>sum_list</tt> with the header format shown below to compute a sum. The function should accept a single input parameter in the form of a list of an arbitrary length containing integer or floating point (decimal) numbers and return a sum for it. The function should be able to handle empty list gracefully by returning 0.    

Test your function for short lists of different length and with different numbers in them. 

In [61]:
def sum_list(x):
  i=0
  sum=0
  for i in range(len(x)):
    x[i] = float(x[i])
    sum += x[i]
  return sum

user_input=input("Enter the elements of the list you want to sum seperated by a space: ")
x=user_input.split()
print("The sum of the numbers in your list is:",sum_list(x))


Enter the elements of the list you want to sum seperated by a space: 3 4 5 
The sum of the numbers in your list is: 12.0


### <font color='blue'>Exercise 3: estimating purity of a candidate quasar sample (5 points)</font> 

**Background info:** astronomers often distinguish two types of sources on the sky by their appearance: diffuse extended sources (such as [galaxies](https://en.wikipedia.org/wiki/Galaxy), nebulae, or comets) and unresolved "point-like" sources (these are the sources that often have ["spikes"](https://www.nasa.gov/sites/default/files/potw1346a.jpg) attached to them in images). The point-like sources are almost all stars in our Galaxy, but a small fraction of them are distant [quasars](https://astronomy.com/news/videos/2021/06/infinity--beyond--episode-20-what-are-quasars). Quasars are interesting for many reasons, but to identify them one needs to sift through hundreds or thousands of stars, so astronomers have developed techniques for classifying point sources into stars and quasars using their colors or some other characteristics. 

This classification is not 100\% accurate, so when it produces a sample of potential quasars, these need to be followed up observationally to get their spectra. This is because unlike their appearance in images, the spectra of quasars and stars are very different, so we can use spectra to definitively identify point sources as either stars or quasars. However, getting spectra is expensive in terms of observational time (and in terms of $\$\$\$$; one night on a large telescope, for example, costs $\sim\$ 100,000$) and human effort. So we are strongly motivated to not waste valuable observing time on candidates that will turn out to be stars. 

**Task 3a. (3 points)** 
Suppose we are conducting a survey to construct a sample of quasars. At the brightness level that we are working at, quasars constitute $1\%$ of point sources. We have a method for classification of point sources into stars and quasars that classifies real stars as stars with probability of 0.99 and real quasars as quasars with probability of 0.99. If the classification method produced 1000 quasar candidates, what fraction of these candidates can we expect to be real quasars (this fraction is called *sample purity* in astronomy)? 

To figure out the fraction of these $1000$ quasar candidates that we can expect to be real quasars, all we need to do is construct a probability tree and utilize Bayes' Theorem and we should be good to go. We'll denote the probability of getting a real quasar given that we classified it as a quasar as $P(Q\mid q)$, where $q$ is classifying an observation as a quasar and $Q$ is it actually being a quasar. We will use the same convention for stars. We can find $P(Q\mid q)$ using Bayes' Theorem:

$$P(Q\mid q)=\frac{P(q\mid Q)\cdot P(Q)}{P(q)}$$

Now we just need to find the probability of classifying an observation as a quasar, $P(q)$, which we can find by simply summing the probability of being initially sorted as a quasar and actually being a quasar, denoted $P(q\text{ and }Q)$, and the probability of being initially sorted as a star and *not* being a star, i.e. being a quasar, which we'll denote $P(q\text{ and }S)$. We also know from basic probability that 

$$P(q\text{ and } Q)=P(Q)\cdot P(q\mid Q)$$ 
and 
$$P(q\text{ and } S)=P(S)\cdot P(q\mid S)$$

So we find that:

$$P(q)=P(q\text{ and } Q)+P(q\text{ and } Q)=P(Q)\cdot P(q\mid Q)+P(S)\cdot P(q\mid S)$$
$$P(q)=0.01(0.99)+0.99(0.01)=0.0198$$

Plugging $P(q)$ into our equation for $P(Q\mid q)$ and using $P(q\mid Q)=0.99$ and $P(Q)=0.01$ from earlier, we find a probability of a classified quasar actually being a quasar of:

$$P(Q\mid q)=\frac{0.99\cdot 0.01}{0.0198}=0.5$$

Implying that in our sample of 1000 quasars, the number of real quasars we would have is:

$$\text{# of Quasars}=1000\cdot 0.5=\boxed{500\text{ Quasars}}$$

**Task 3b (2 points).** Vary the fraction of quasars among point sources and the probabilities for classifier to label real stars as stars and real quasars as quasars. Which of these probabilities affect the sample purity the most?

To determine which of the probabilities affects the sample purity the most, we'll vary each of the probabilities by halving them and observing the resultant effect on the probability of a quasar actually being a quasar, $P(Q\mid q)$. Instead of going one by one and adjusting the variables by one increment at a time, I threw the equation from part (a) for $P(Q\mid q)$ into Desmos with variable sliders: $s$ on the interval 0 to 0.1 with steps of 0.005, where s is the probability of a real star being classified as a quasar (couldn't write $P(q\mid S)$ in Desmos, so had to abbreviate); $q$ on the interval 0.9 to 1 with steps of 0.005, where q is the probability of a real quasar being classified as a quasar; and $Q$ on the interval 0 to 0.1 with steps of 0.005, where Q is the fraction of quasars amongst point sources. Starting off from our baseline in part (a) with $s=0.01$, $q=0.99$, and $Q=0.01$, I noticed that $q$ probably had the least effect. Sliding through the values of $q$ on this interval barely affected the value for $P(Q\mid q)$, it changed $P(Q\mid q)$ by maybe $0.04$ across its interval. Meanwhile, sliding through $s$ had a more dramatic effect, changing the value for $P(Q\mid q)$ by nearly 0.5 across the its own interval. However, I think $Q$ had the biggest effect, as when I slid through the interval I placed it on, $P(Q\mid q)$ changed by over 0.6 across its respective interval. So I think $Q$, the fraction of quasars amongst point sources, has the biggest affect on the sample purity, but $s$, the probability of a real star being classified as a quasar was definitely not a far second in terms of effects, especially compared to how minimal of an effect $q$, the probability of a real quasar being classified as a quasar, had on our value for $P(Q\mid q)$.

### <font color='blue'>Exercise 4: testing isotropy of the Universe (8 points)</font> 

Using the sample of quasars we constructed after much effort (detailed in the previous exercise and after we obtained spectra and removed contaminating stars), we measured that the average density of quasars on the sky in the sample is 5 quasars per square degree.

You may have heard that our Universe is often assumed to be *isotropic* on largest scales (i.e. has the same properties  in all directions), but when we were analyzing one of the sky regions of one square degree area in your survey, we have noticed that there are 11 quasars in that region.   

**_Note_**: A square degree is the area of a square region on the sky of 1 degree of arc on the sky - one could fit approximately 4 full Moons into such a square.


**Task 4a. (4 points)** Estimate the probability to observe *at least* 11 quasars in that region. Comment on whether the probability is low or high. 

We can estimate the probability to observe at least 11 quasars in that region by finding the probability of observing 10 or less quasars in that region and subtracting our result from 1. Writing a simple program for that using our myfactorial program from last homework, we find:

In [62]:
def myfactorial(n):
  if n<0:
    print("Invalid input! Please rerun the program and enter a non-negative integer.")
  elif n==0:
    return 1
  else:
    for i in range(1,n):
      if i<n:
        n *= i
    return n

In [63]:
from math import e

def ppoisson(n,lam):
  p=0
  p=(lam**(n))/myfactorial(n)*e**(-lam)
  return p

n=0
P=0

for n in range (0,11):
  P += ppoisson(n,5)
print(1-P)

0.01369526859838266


So the probability of observing at least 11 quasars in that region is $\boxed{P\approx 0.0137}$ .

**Task 4b. (4 points)** How big should the area of our survey be approximately in order for this concentration of 11 quasars to be consistent with a uniform/isotropic distribution of quasars on the sky? Present quantitative argument(s) to support your result. 

In order to determine the area of our survey required for this concentration of 11 quasars to be consistent with an isotropic distribution of quasars in the sky, we can find the area in which we would expect to find at least one of these distributions of quasars per square degree. Since we have the frequency $P=f=0.0137$ for observing at least 11 quasars in a region, we can find the period $T$ using the simple relation:

$$T=\frac{1}{f}$$

Plugging in a frequency of $f=0.0137$ for such regions, we find a period $T$ of:

$$T=\frac{1}{0.0137}=73.0\text{ square degrees}$$

Implying that, within an area of $73.0$ square degrees in the sky, we can expect to find at least one such region containing at least 11 quasars. That is, we would need a survey area of $\boxed{73.0\text{ square degrees}}$ in order to be consistent with an isotropic distribution of quasars on the sky.

### <font color='blue'>Exercise 5: have we detected a source? (8 points)</font> 

**Background information.** Detailed discussion of the source and background pixels in CCD images and how photon counts can be used to make conclusions about statistical *significance* of a given potential source (i.e., how likely that it is real) can be found in the [03_probability_distributions_poisson](https://drive.google.com/file/d/12vmgU44wLBNsaCbQ5EVjc3Ta0dMczupt/view?usp=sharing) notebook. It may be helpful to review it before thinking about how to approach this exercise. 


You are hunting for the most distant galaxy known and find a possible source in a deep sky image.  
A CCD detector receives 1400 photons in pixels corresponding to a sky region containing the possible source.
1100 photons are received on average in the same number of "background" pixels and the same exposure time that do not receive photons from the source. These photons are due to a sky background that is approximately uniform on the sky and in time.  

**Task 5a. (3 points)**
Is it time to start writing discovery paper? In other words, can you present compelling arguments that it is likely to be a real source and not a statistical fluctuation in the sky background? The argument should include the list of assumptions being made in your assessment and a list of possible complicating effects that cannot be addressed with
the given information. 




In order to determine if what we're seeing is a real source and not just a statistical fluctuation in the sky background, we can calculate the Signal-to-Noise Ratio (SNR) and check if the value is $\geq 5$, which is usually when a source can begin to be considered real. Given the equation for the Signal-to-Noise ratio:

$$\mathrm{SNR}=\frac{n_{\mathrm{tot}}-n_{b}}{\sqrt{n_{\mathrm{tot}}+n_{b}}}$$

Where $n_{\mathrm{tot}}$ is the total number of photons detected and $n_{\mathrm{b}}$ is the number of background photons associated with the source. We are given both these values in the question prompt, specifically $n_{\mathrm{tot}}=1400$ and $n_{\mathrm{b}}=1100$, so plugging these values into the equation above, we find a Signal-to-Noise Ratio SNR of:

$$\mathrm{SNR}=\frac{1400-1100}{\sqrt{1400+1100}}=\frac{300}{50}=6$$

Since we have found a Signal-to-Noise Ratio of $\mathrm{SNR}=6$, it is statistically safe to say that the source *may* be considered real. However, there are a few other considerations we would have to take into account before we begin writing our discovery paper. First, it may be smart to take another few samples of photons from this region, as we do not know if this spike in the number of photons is actually due to the detection of a possible distant galaxy, or it could be from some sort of interference of human origin (i.e. airplane, satellite, light pollution). We would have to make quite a few assumptions about the night we observed this spike in photons (namely, the ones just mentioned -- no light pollution, no interference, sufficient sample size, etc.), many of which we cannot make with the given information  and/or the data from this one night.

**Task 5b. (5 points)**
To make sure your measurements from the image mentioned above are robust, you apply for time and get another image of this sky region using the same telescope and exposure time. In this image you measure an average of 1150 photons in the   blank regions away from the source in the same number of pixels as the source (i.e. background pixels). Now you have two measurements of the sky background: 1100 and 1150 photons. 

Does this new measurement weaken your conclusion from 1a or does it help strengthen it? 

To check if this new measurement weakens our conclusion from part (a) or strengthens it, we'll compute the new Signal-to-Noise Ratio but instead incorporating our second measurement of $n_{b2}=1150\text{ photons}$ into the mix. We can do this by finding the mean of our background photon measurements:

$$\overline{n_b}=\frac{n_{b1}+n_{b2}}{2}=\frac{1100+1150}{2}=1125\text{ photons}$$

So that goes in the numerator. To adjust our denominator to account for the additional data point, we can recall that the variance of a Poisson distribution is its mean. Using the linearity of the variance and accounting for the square root, we have that:

$$\sigma^2_{nb}=\sigma^2\left(\frac{n_{b1}+n_{b2}}{2}\right)=\frac14(n_{b1}+n_{b2})$$

Plugging this into our equation for the Signal-to-Noise Ratio from part (a) and using $\overline{n_b}=1125\text{ photons}$ as our value for $n_b$, we find:

$$\mathrm{SNR}=\frac{n_{\mathrm{tot}}-\overline{n_b}}{\sqrt{n_{\mathrm{tot}}+\frac14(n_{b1}+n_{b2})}}=\frac{1400-1125}{\sqrt{1400+\frac14(1100+1150)}}=6.208$$

So our Signal-to-Noise Ratio for this new data set is $\boxed{\mathrm{SNR}\approx 6.23}$ , which is greater than our previous $\mathrm{SNR}$ value of $6$ in part (a), implying that adding this additional background photon number data point **strengthened** our conclusion from part (a).

### <font size=4 color='blue'>Exercise 6: implementing a flexible sum function (8 points + 4 extra-credit)</font> 

Implement a flexible function to compute a sum that (options increase in difficulty from top to bottom) 

**6a (4 points)** can compute the sum for an arbitrary number of input integers or float numbers. 
In other words, calls like this should all work: 

    flexible_sum(x1, x2) # sum of two integer or float numbers or strings x1 and x2
    flexible_sum(x1, x2, x3) # sum of 3 input parameters 
    
**6b (2 points)** above *and* can also handle a single input parameter, returning it unchanged and with the same type, i.e.: 
   
       flexible_sum(1)
       
   returns 1 
   
**6c (2 points)** above *and* can accept an arbitrary number of strings as input parameters and return their concatenation. In other words, the following call
   
       flexible_sum('one ', 'two ', 'three ')
       
   should give result string
   
       'one two three '

**6d. extra-credit (4 points)**: above, but the function can also accept a single input parameter in the form of a list of an arbitrary length and return a sum for it. 

If you can, try to implement your function without calls to any other external functions, including Python built-in functions. 
